src/AdoPat/PatClient.cs

```diff
@@ -0,0 +1,139 @@
+            do
+            {
+                pagedPatTokens = await this.client.ListPatsAsync(
+                    displayFilterOption: DisplayFilterOptions.Active,
+                    top: PageSize,
+                    continuationToken: pagedPatTokens.ContinuationToken,
+                    sortByOption: SortByOptions.DisplayDate,
+                    cancellationToken: cancellationToken)
+                .ConfigureAwait(false);
+
+                foreach (PatToken pat in pagedPatTokens.PatTokens)
+                {
+                    pats.Add(pat);
+                }
+            }
+            while (!string.IsNullOrEmpty(pagedPatTokens.ContinuationToken));
+
+            return pats;
+        }
+
+        /// <summary>
+        /// Creates a new PAT with a new 'valid to' date by replicating an existing one.
+        /// </summary>
+        /// <param name="patToken">An existing <see cref="PatToken"/>.</param>
+        /// <param name="validTo">The new expiration date for the regenerated token.</param>
+        /// <param name="cancellationToken">The cancellation token to cancel the operation.</param>
+        /// <returns>A <see cref="Task"/> returning a <see cref="PatToken"/>.</returns>
+        public async Task<PatToken> RegeneratePatAsync(
+            PatToken patToken,
+            DateTime validTo,
+            CancellationToken cancellationToken = default)
+        {
+            // Regeneration is an option for existing PATs in the Azure DevOps UI,
+            // but from the API's viewpoint it is an abstraction over 3 operations.
+            //
+            // 1. Understand the metadata of the PAT using a GET call.
+            // 2. Create a new PAT with the old PAT's metadata using a POST call.
+            // 3. Revoke the old PAT using a DELETE call.
+            //
+            // We skip the initial GET call because the given patToken should have
+            // its own metadata. It risks being out of date depending on where it
+            // was fetched from.
+            //
+            // For more info see:
+            // https://learn.microsoft.com/en-us/azure/devops/organizations/accounts/manage-personal-access-tokens-via-api?view=azure-devops#q-how-can-i-regeneraterotate-pats-through-the-api-i-saw-that-option-in-the-ui-but-i-dont-see-a-similar-method-in-the-api
+            PatTokenCreateRequest patTokenCreateRequest = new PatTokenCreateRequest
+            {
+                DisplayName = patToken.DisplayName,
+                Scope = patToken.Scope,
+                ValidTo = validTo,
+                AllOrgs = AllOrgs,
+            };
+
+            var renewedPatToken = await this.CreatePatAsync(
+                patTokenCreateRequest,
+                cancellationToken: cancellationToken)
+            .ConfigureAwait(false);
+
+            // XXX: If we create a new PAT, but fail to revoke the old PAT, should
+            // we continue or should we try to undo the creation and bail? It's

```

Comment:
Had some questions to understand the scenario better
1. If the token has been revoked, we would still be able to create a new PAT using it's metadata - is that correct?
2. If the old PAT has been revoked and if we create a new PAT, there can be 2 scenarios if I understand correctly
a. User has already created a new PAT ()with same name) when they revoked previous which we won't know about?
b. There is no active PAT and we just create a new one.
While for scenario `b`, the user would be in good state if we continue, I am curious how we will handle (if at all it's possible) scenario `a`.?
3. How do we undo creation? Will we send another `revoke` request? While this might be addressed later so feel free to not answer - do we clean up the cache?

Reply:
1. Yes, if a token has been revoked, but we still have its metadata it should be possible to create a new one with the same metadata.
2. For scenario `b`, yes, if there's no active PAT which corresponds to the one that was just requested, we can keep the new one. I'm not sure that I understand scenario `a` here, can you try to explain what you mean? 🤔
3. Yes, unfortunately there's no such thing as deleting a PAT. Creation can only be "undone" by sending another `revoke` request. If we did "undo" I think it would be wise to clean up the cache. In this case we won't necessarily have written the PAT we _just_ created to the cache, but we can remove the old entry (if one exists).
4. I'm not sure. I don't have certainty on the exact frequency, but I have only _observed_ emails being sent upon creation of a new PAT.

Reply:
Thanks for answering all those question :)
For scenario `a`, I was wondering that if the user has used UI to regenerate the PAT - that would make old PAT invalid. Now, when we create new a PAT (because old became invalid when user used UI to regenerate), would it interfere with the previously (UI) regenerated PAT which is also active? Or would both of them work independently (meaning we will not bother about the PAT regenerated through UI and just create new one -> store it in cache and remove any invalid)?

I think we would not be bothered about (UI) regenerated PAT and just regenerate again, but just wanted to confirm.

Reply:
The user using the UI to create a new PAT doesn't necessarily invalidate an older PAT. You can have more than one PAT with the same name and scopes. Unless the user themselves marks it as revoked or the PAT expires it's probably still valid. The UI created PAT would continue to work independently. At least as far as I understand Azure DevOps PATs.

My inclination here is to just ignore PATs that aren't in our cache even if we could discover through the API that they've independently created a PAT which matches the name/scopes we're looking for. I'm especially inclined to do so because we can never get the token value for that PAT, so it's useless to the machine running our CLI.